src/core/ComboBox/ComboBox.tsx

```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
     }
-    return filteredOptions.map((option) => {
-      const index = options.findIndex(({ value }) => option.value === value);
-      if (index < 0) {
-        return <></>;
-      }
+    case 'close': {
+      return { ...state, isOpen: false };
+    }
+    case 'toggle': {
+      return { ...state, isOpen: !state.isOpen };
+    }
+    case 'select': {
+      return {
+        ...state,
+        isOpen: false,
+        selectedIndex: value ?? state.focusedIndex ?? state.selectedIndex,
+        focusedIndex: value ?? state.focusedIndex,
+      };
+    }
+    case 'focus': {
+      return { ...state, focusedIndex: value ?? state.selectedIndex ?? -1 };
+    }
+    default: {
+      return state;
+    }
+  }
+};
 
-      const id = getOptionId(index);
-      const isSelected = selectedValue === option.value;
-      const isFocused = focusedIndex === index;
-      const focusScrollRef = (el: HTMLElement) =>
-        isFocused && el?.scrollIntoView({ block: 'nearest' });
-
-      if (isSelected || isFocused) {
-        const item =
-          itemRenderer?.(option, { index, id, isSelected, isFocused }) ??
-          React.cloneElement(memoizedItems[index], { isSelected });
-
-        return React.cloneElement(item, {
-          className: cx({ 'iui-focused': isFocused }, item.props.className),
-          ref: mergeRefs(focusScrollRef, item.props.ref),
-          value: option.value,
-          role: 'option',
-          onClick: () => {
-            setSelectedValue(option.value);
-            userOnChange.current?.(option.value);
-            setIsOpen(false);
-          },
-        });
-      }
+const ComboBoxRefsContext = React.createContext<
+  | {
+      inputRef: React.RefObject<HTMLInputElement>;
+      menuRef: React.RefObject<HTMLUListElement>;
+      toggleButtonRef: React.RefObject<HTMLSpanElement>;
+      optionsRef: React.MutableRefObject<Array<HTMLLIElement>>;
+    }
+  | undefined
+>(undefined);
+ComboBoxRefsContext.displayName = 'ComboBoxRefsContext';
 
-      return memoizedItems[index];
-    });
-  }, [
-    filteredOptions,
-    emptyStateMessage,
-    options,
-    getOptionId,
-    selectedValue,
-    focusedIndex,
-    itemRenderer,
-    memoizedItems,
-  ]);
+const ComboBoxStateContext = React.createContext<
+  | {
+      isOpen: boolean;
+      id: string;
+      minWidth: number;
+      focusedIndex?: number;
+    }
+  | undefined
+>(undefined);
+ComboBoxStateContext.displayName = 'ComboBoxStateContext';
+
+const ComboBoxActionContext = React.createContext<
+  ((x: [ComboBoxAction] | [ComboBoxAction, number]) => void) | undefined
+>(undefined);
+ComboBoxActionContext.displayName = 'ComboBoxActionContext';
+
+const useSafeContext = <T,>(context: React.Context<T>) => {
+  const value = React.useContext(context);
+  if (!value) {
+    throw new Error(`${context.displayName} is undefined`);
+  }
+  return value!; // eslint-disable-line @typescript-eslint/no-non-null-assertion -- we already checked for undefined
```

Comment:
this is weird.. why it fails to see the check above? because no return?
Maybe this whole function can be moved to utils and reused, right?